### ISSUE TITLE
[Ecosystem] RLinf

### ISSUE THREAD

**Issue by andylin-hao** [pytorch-fdn/ecosystem#70]
### Contact emails

- [linhao@infini-ai.com](mailto:linhao@infini-ai.com) Hao Lin
- [guozhen@infini-ai.com](mailto:guozhen@infini-ai.com) Zhen Guo
- [zhangquanlu@infini-ai.com](mailto:zhangquanlu@infini-ai.com) Quanlu Zhang
- [zoeyuchao@gmail.com](mailto:zoeyuchao@gmail.com) Yu Chao

### Project summary

Reinforcement learning library for embodied and agentic AI.

### Project description

RLinf is a flexible and scalable reinforcement learning infrastructure for embodied and agentic AI.

RLinf is flexible and research-friendly:

- the Macro-to-Micro Flow (M2Flow) paradigm decouples high-level RL workflows from low-level execution, communication, and scheduling
- it supports PPO, GRPO, DAPO, Reinforce++, SAC, IQL, DAgger, SFT, VLA fine-tuning, real-world robot learning, math reasoning RL, search agents, coding online RL, and multi-agent RL
- it integrates with models, simulators, and robotics platforms including OpenVLA, pi0/pi0.5, GR00T, ManiSkill, LIBERO, IsaacLab, RoboTwin, Franka, and DOS-W1

RLinf is fast and production-ready:

- configurable placement supports collocated, disaggregated, and hybrid execution across actor, rollout, environment, reward, and agent workers
- FSDP + Hugging Face supports fast prototyping, while Megatron + SGLang/vLLM supports large-scale training and rollout
- hybrid embodied RL execution is documented as achieving up to 2.434x throughput over existing frameworks

RLinf contributes to the PyTorch community:

- it provides a practical RL systems testbed for PyTorch FSDP/FSDP2, DTensor, torch.distributed, torch.distributed.checkpoint, torch.compile, CUDA graphs, and mixed precision
- its adaptive asynchronous Channel uses optimized backends such as GLOO and NCCL to overlap communication with rollout and training
- its scheduler extends PyTorch-based distributed training to heterogeneous hardware, including node groups, accelerator ranks, CPU-only nodes, and robot hardware

### Are there any other projects in the PyTorch Ecosystem similar to yours?  If, yes, what are they?

Yes. Related projects include TorchRL and verl.

RLinf is complementary to these projects. Compared with general RL libraries, RLinf focuses on scalable infrastructure for embodied and agentic AI, including VLA/robotics training, real-world online policy learning, world-model-based RL, and heterogeneous distributed execution. Compared with LLM-focused RL training systems, RLinf covers both agentic LLM post-training and embodied AI workflows, with dedicated environment workers, simulator integrations, robotics examples, real-world deployment support, and configurable component placement across actor, rollout, environment, reward, and agent components.

### Project repo URL

https://github.com/RLinf/RLinf

### Additional repos in scope of the application

None.

### Project license

Apache-2.0 license

### GitHub handles of the project maintainer(s)

- [@andylin-hao](https://github.com/andylin-hao)
- [@guozhen1997](https://github.com/guozhen1997)
- [@zoeyuchao](https://github.com/zoeyuchao)
- [@QuanluZhang](https://github.com/QuanluZhang)

### Is there a corporate or academic entity backing this project?  If so, please provide the name and URL of the entity.

Tsinghua University: https://www.tsinghua.edu.cn/en/, Infinigence AI: https://www.infini-ai.com/

### Website URL

https://rlinf.readthedocs.io/

### Documentation

Link: https://rlinf.readthedocs.io/

Documentation table of contents:

- Start
  - Installation
  - VLA quickstart
  - VLA evaluation
  - LLM evaluation
- Tutorials
  - User configuration and YAML
  - Execution modes: collocated, disaggregated, and hybrid
  - Advanced topics including logging, checkpoint resume, heterogeneous clusters, and weight synchronization
  - Extending RLinf with new algorithms, models, environments, and reward models
  - RL algorithms including PPO, GRPO, DAPO, Reinforce++, SAC, CrossQ, RLPD, and IQL
- Examples
  - Embodied AI examples for ManiSkill, LIBERO, IsaacLab, RoboTwin, RoboVerse, BEHAVIOR, MetaWorld, CALVIN, RoboCasa, Franka, XSquare Turtle2, DOS-W1, and other simulators or real-world platforms
  - Agentic AI examples including SearchR1, rStar2, online coding RL, math reasoning RL, and WideSeek-R1
  - SFT and VLA fine-tuning examples
  - System examples such as FUSCO
- Blog
- Publications
- API references
- FAQ

### How do you build and test the project today (continuous integration)?  Please describe.

RLinf is a Python project published as a package named `rlinf` and installable through PyPI for library use. For full experiments, RLinf provides Docker images and `requirements/install.sh` targets for embodied, agentic/reasoning, and documentation environments.

The repository uses GitHub Actions for continuous integration. Current workflows include:

- Unit tests on CPU and CUDA runners, including `pytest tests/unit_tests`, scheduler doctests, package build tests, package import tests, and package structure checks.
- Scheduler tests for static auto placement and dynamic scheduler scenarios.
- Installation tests for agentic/reasoning and embodied model/environment combinations.
- Docker build tests for reasoning and embodied images.
- End-to-end embodied RL tests, including ManiSkill SAC, asynchronous SAC, SAC-Flow, D4RL offline IQL, and other embodied workflows.
- End-to-end agent/reasoning tests covering Megatron and FSDP training with SGLang and vLLM in collocated and pipeline modes.
- SFT end-to-end tests for VLA and related workflows.
- Lint, commit-message, and PR-title checks.

The main CI entrypoint gates pull requests to `main` and release branches.

### Version of PyTorch

The package currently supports `torch>=2.5.0,<=2.9.0`.

### Components of PyTorch

RLinf uses and integrates with several PyTorch components and PyTorch-adjacent distributed training features:

- PyTorch FSDP and FSDP2 for model training and sharding.
- PyTorch DTensor in FSDP actor and utility paths.
- Pytorch distributed communication primitives such as GLOO, NCCL, and IPC-based multiprocess communication.
- PyTorch native `nn.Module` support through the FSDP + Hugging Face training path.
- PyTorch distributed training primitives as part of RLinf's actor, worker, and backend integrations.
- `torch.compile` support in rollout/SGLang configuration paths and runtime performance optimizations.
- `torchdata` as part of the package dependency set.
- CUDA and mixed-precision training support through PyTorch-based actor and model manager implementations.

### How long do you expect to maintain the project?

Indefinitely.

### Additional information

- Repository: https://github.com/RLinf/RLinf
- Documentation: https://rlinf.readthedocs.io/
- PyPI: https://pypi.org/project/rlinf/
- Hugging Face: https://huggingface.co/RLinf

Community links:

- Governance: https://github.com/RLinf/governance
- Steering committee: https://github.com/RLinf/governance/blob/main/STEERING-COMMITTEE.md
- Contributing guide: https://github.com/RLinf/RLinf/blob/main/CONTRIBUTING.md
- Code of conduct: https://github.com/RLinf/RLinf/blob/main/CODE_OF_CONDUCT.md
- Code owners: https://github.com/RLinf/RLinf/blob/main/.github/CODEOWNERS

Documentation and publication links:

- Example gallery: https://rlinf.readthedocs.io/en/latest/rst_source/examples/index.html
- API references: https://rlinf.readthedocs.io/en/latest/rst_source/apis/index.html
- FAQ: https://rlinf.readthedocs.io/en/latest/rst_source/faq.html
- Main technical report: https://arxiv.org/abs/2509.15965

**Comment by hogepodge** [comment]
Hi @andylin-hao I'll get started on the project review tomorrow afternoon. Thanks for your patience!

**Comment by hogepodge** [comment]
Title | Description | RLinf | RLinf Notes
-- | -- | -- | --
PyTorch Support | Project must demonstrate benefit to the pytorch community and be validated with the latest versions of PyTorch. | ⚠️ Partial |  PyTorch support limited to <= 2.9.0
Working CI | Projects must have a demonstrable working CI workflow for all declared architectures | ✅ Yes |  
Governance | Projects must have documented technical governance defined and implemented and it must be represented in their Github or Gitlab repository. Example: Review the LF's Minimum Viable Governance framework. | ✅ Yes | Minimum viable governance
User experience | The value proposition should be made clear for the project and documented in the README of the project repository. | ✅ Yes |  
Quick start | Project should have a working and demonstrable quick start. | ✅ Yes |  
Documentation | Projects must have clear and comprehensive documentation. | ✅ Yes |  
Permissive licensing | Users must be able to utilize ecosystem projects without licensing concerns. e.g. BSD-3, Apache-2 and MIT licenses | ✅ Yes |  
Functional Testing | Users should have developed tests submitted as a part of their project, and be confident that the project will work well with the latest releases of PyTorch. | ✅ Yes | e2e and unit tests
Packages | Projects need to have support for binary installation options (pip/Conda). | ✅ Yes | pip and Docker packages
Ongoing maintenance | Project maintainers must be committed to supporting and maintaining their projects for the next 12 months at a minimum. Projects must commit to updating their projects to the latest two releases of PyTorch (e.g. versions 2.6, 2.5.1). | ✅ Yes |  
Core Maintainers | The project must have a minimum of 2 core maintainers | ✅ Yes | 
Contributors | The project should have a minimum of 5 total contributors active in the last 90 days. | ✅ Yes |  
Commits | The project must have 30 commits in the last 90 days. | ✅ Yes |  
Stars | The project must have a minimum of 200 stars on github | ✅ Yes | 3400 stars
LICENSE.md | at the root of the repository specifying the terms and conditions for using, distributing, and modifying the software. In addition, you should provide information on the license of any third-party code included in the project. | ✅ Yes | Apache 2.0
README.md | welcomes new community members to the project and explains why the project is useful and how to get started. Should include information about release methodology, cadence, and criteria in the README. | ✅ Yes |  
CONTRIBUTING.md | explains how to contribute to the project. The file explains the types of contributions needed and how the development process works. | ✅ Yes |  
CODEOWNERS | that define individuals or teams responsible for code in a repository; document current project owners and Retired committers. | ✅ Yes |  
CODE_OF_CONDUCT.md | sets the ground rules for participants' behavior and helps facilitate a friendly, welcoming environment. By default, projects should leverage the Linux Foundation Code of Conduct unless an alternate Code of Conduct is approved prior. | ✅ Yes | Contribuor Covenant 2.0

**Comment by hogepodge** [comment]
@andylin-hao thanks for the submission. The only major gap I see in the requirements is a lack of governance documentation.

**Comment by andylin-hao** [comment]
> [@andylin-hao](https://github.com/andylin-hao) thanks for the submission. The only major gap I see in the requirements is a lack of governance documentation.

Hi @hogepodge, we use Minimum Viable Governance. The doc is here https://github.com/RLinf/governance. Would this suffice?

**Comment by andylin-hao** [comment]
> > [@andylin-hao](https://github.com/andylin-hao) thanks for the submission. The only major gap I see in the requirements is a lack of governance documentation.
> 
> Hi [@hogepodge](https://github.com/hogepodge), we use Minimum Viable Governance. The doc is here https://github.com/RLinf/governance. Would this suffice?

@hogepodge Kind ping~

**Comment by hogepodge** [comment]
Yes, Minimum Viable Governance is great. The requirement is there to ensure that project owners have given consideration to a governance structure, and this meets the requirement. Thank you!

**Comment by michellearoth** [comment]
Accepted! 

<img width="715" height="954" alt="Image" src="https://github.com/user-attachments/assets/42cd4b55-d77e-4ea9-8e8e-2e44b47a2a6a" />

**Comment by andylin-hao** [comment]
> Accepted!
> <img alt="Image" width="715" height="954" src="https://private-user-images.githubusercontent.com/157139387/598920545-42cd4b55-d77e-4ea9-8e8e-2e44b47a2a6a.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Nzk5MDA3OTcsIm5iZiI6MTc3OTkwMDQ5NywicGF0aCI6Ii8xNTcxMzkzODcvNTk4OTIwNTQ1LTQyY2Q0YjU1LWQ3N2UtNGVhOS04ZThlLTJlNDRiNDdhMmE2YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwNTI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDUyN1QxNjQ4MTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04N2Y5NmZkYWI2MmQ3MzViN2JkN2RhZmMwZGY3ZDcxNzA4Mzg2Y2VlYzA5NWJjYTM4ZDJjN2U5YTI5NThkZDk4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.UwnxSMRVLIxSvNXJYpGM935WHb9WOJfV9jxMzzhHkpU">

Thanks! Good to hear!

**Comment by hogepodge** [comment]
Congratulations @andylin-hao!
